spring-integration-core/src/main/java/org/springframework/integration/aop/AbstractMessageSourceAdvice.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.aop;
+
+import org.aopalliance.intercept.MethodInterceptor;
+import org.aopalliance.intercept.MethodInvocation;
+
+import org.springframework.integration.core.MessageSource;
+import org.springframework.messaging.Message;
+
+/**
+ * Advice for a {@link MessageSource#receive()} method to decide whether a poll
+ * should be ignored and/or take action after the receive.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public abstract class AbstractMessageSourceAdvice implements MethodInterceptor {
+
+	@Override
+	public final Object invoke(MethodInvocation invocation) throws Throwable {
+		Object target = invocation.getThis();
+		if (!(target instanceof MessageSource)
+				|| invocation.getMethod().getName() != "receive") {
+			return invocation.proceed();
+		}
+
+		Message<?> result = null;
+		if (beforeReceive((MessageSource<?>) target)) {
+			result = (Message<?>) invocation.proceed();
+		}
+		return afterReceive(result, (MessageSource<?>) target);
+	}
+
+	/**
+	 * Subclasses can decide whether to proceed with this poll.
+	 * @param target the message source.
+	 * @return true to proceed.
+	 */
+	public abstract boolean beforeReceive(MessageSource<?> source);
```

Comment:
I'm still not sure in this feature. We should keep in mind that we can have `max-messages-per-poll > 1`, hence any changes of this advice for the current `receive()` won't affect any other `receive()` during the same `poll()`. Hence we should be a bit careful with `Trigger` adjusting on each `receive()`. See `ReschedulingRunnable`: any changes to the trigger will be accepted only after the finish of the current `poll()`.

Maybe we must consider `Future.cencel(true)` as well ? But how we reschedule it afterward...
I see you wrap `MessageSource` to the proxy in the `SourcePollingChannelAdapter`. Should this method (and the next one) accept the `target` instead of the current proxy? Or does `invocation.getThis()` return exactly the `target` ?